flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java

```diff
@@ -54,117 +75,494 @@
+			throw new CatalogException(
+				String.format("Failed to get table %s from Hive metastore", tablePath.getFullName()), e);
+		}
+	}
+
+	private static CatalogBaseTable instantiateHiveCatalogTable(Table hiveTable) {
+		boolean isView = TableType.valueOf(hiveTable.getTableType()) == TableType.VIRTUAL_VIEW;
 
 		// Table properties
 		Map<String, String> properties = hiveTable.getParameters();
-
-		// Table comment
+		boolean isGeneric = Boolean.valueOf(properties.get(GENERC_META_PROPERTY_KEY));
+		if (isGeneric) {
+			properties = retrieveFlinkProperties(properties);
+		}
 		String comment = properties.remove(HiveTableConfig.TABLE_COMMENT);
 
+		// Table schema
+		TableSchema tableSchema =
+			HiveTableUtil.createTableSchema(hiveTable.getSd().getCols(), hiveTable.getPartitionKeys());
+
 		// Partition keys
 		List<String> partitionKeys = new ArrayList<>();
-
 		if (!hiveTable.getPartitionKeys().isEmpty()) {
-			partitionKeys = hiveTable.getPartitionKeys().stream()
-				.map(fs -> fs.getName())
-				.collect(Collectors.toList());
+			partitionKeys = hiveTable.getPartitionKeys().stream().map(fs -> fs.getName()).collect(Collectors.toList());
 		}
 
-		if (TableType.valueOf(hiveTable.getTableType()) == TableType.VIRTUAL_VIEW) {
-			return new HiveCatalogView(
-				hiveTable.getViewOriginalText(),
-				hiveTable.getViewExpandedText(),
-				tableSchema,
-				properties,
-				comment
-			);
+		if (isView) {
+			if (isGeneric) {
+				return new GenericCatalogView(
+					hiveTable.getViewOriginalText(),
+					hiveTable.getViewExpandedText(),
+					tableSchema,
+					properties,
+					comment
+				);
+			} else {
+				return new HiveCatalogView(
+					hiveTable.getViewOriginalText(),
+					hiveTable.getViewExpandedText(),
+					tableSchema,
+					properties,
+					comment
+				);
+			}
 		} else {
-			return new HiveCatalogTable(
-				tableSchema, partitionKeys, properties, comment);
+			if (isGeneric) {
+				return new GenericCatalogTable(tableSchema, partitionKeys, properties, comment);
+			} else {
+				return new HiveCatalogTable(tableSchema, partitionKeys, properties, comment);
+			}
 		}
 	}
 
-	@Override
-	protected Table createHiveTable(ObjectPath tablePath, CatalogBaseTable table) {
-		Map<String, String> properties = new HashMap<>(table.getProperties());
-
-		// Table comment
-		properties.put(HiveTableConfig.TABLE_COMMENT, table.getComment());
-
+	private  static Table instantiateHiveTable(ObjectPath tablePath, CatalogBaseTable table) {
 		Table hiveTable = new Table();
 		hiveTable.setDbName(tablePath.getDatabaseName());
 		hiveTable.setTableName(tablePath.getObjectName());
 		hiveTable.setCreateTime((int) (System.currentTimeMillis() / 1000));
 
+		Map<String, String> properties = new HashMap<>(table.getProperties());
+		// Table comment
+		properties.put(HiveTableConfig.TABLE_COMMENT, table.getComment());
+		if (table instanceof GenericCatalogTable || table instanceof GenericCatalogView) {
+			properties = maskFlinkProperties(properties);
```

Comment:
do we need to add the "flink.is_generic" key?

Reply:
THE key is added in each of meta class (view, tablle, etc).